File: solutions/number-of-enclaves/number-of-enclaves.cpp
// @Title: Number of Enclaves
// @Author: colinjxc
// @Date: 2022-02-12T18:01:38+08:00
// @URL: https://leetcode-cn.com/problems/number-of-enclaves


class Solution {
public:
    int numEnclaves(vector<vector<int>>& grid) {
      int cnt = 0, m = grid.size(), n = grid.front().size();
      for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
          cnt += grid[i][j];
        }
      }

      function<void(int,int)> dfsSubtract = [&](int i, int j) {
        if (i < 0 || j < 0 || i == m || j == n || grid[i][j] == 0) {
          return;
        }
        grid[i][j] = 0;
        cnt--;
        dfsSubtract(i, j-1);
        dfsSubtract(i-1, j);
        dfsSubtract(i, j+1);
        dfsSubtract(i+1, j);
      };

      for (int i = 0; i < m; i++) {
        dfsSubtract(i, 0);
        dfsSubtract(i, n-1);
      }

      for (int i = 0; i < n; i++) {
        dfsSubtract(0, i);
        dfsSubtract(m - 1, i);
      }
      return cnt;
    }
};

File: solutions/number-of-enclaves/number-of-enclaves.go
// @Title: Number of Enclaves
// @Author: colinjxc
// @Date: 2022-02-12T18:39:46+08:00
// @URL: https://leetcode-cn.com/problems/number-of-enclaves


func numEnclaves(grid [][]int) int {
    m, n := len(grid), len(grid[0])
    cnt1 := 0
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            cnt1 += grid[i][j]
        }
    }
    var dfs func (int, int)
    dfs = func(i, j int) {
        if i < 0 || j < 0 || i == m || j == n || grid[i][j] == 0 {
            return
        }
        grid[i][j] = 0
        cnt1--
        dfs(i, j - 1)
        dfs(i - 1, j)
        dfs(i, j + 1)
        dfs(i + 1, j)
    }

    for i := 0; i < m; i++ {
        dfs(i, 0)
        dfs(i, n - 1)
    }

    for i := 0; i < n; i++ {
        dfs(0, i)
        dfs(m - 1, i)
    }
    return cnt1;
}
